[TLS 1.3] Support Session Resumption for TLS Server

reneme opened the pull request:
## Pull Request Dependencies

* ~~https://github.com/randombit/botan/pull/3053~~
* ~~https://github.com/randombit/botan/pull/3244~~
* ~~https://github.com/randombit/botan/pull/3150~~

## Description

This adds facilities for PSK-based session resumption to the TLS 1.3 server implementation. The resumption mechanism of TLS 1.3 differs substantially from previous protocol versions: Therefore, this required changes in the `Session_Manager` API that can be found in #3150.

### New Public API

#### `TLS::Server::send_new_session_tickets(const size_t tickets)`

When called on TLS 1.3 server channel that finished its handshake, this sends "`tickets`" [`NewSessionTicket` messages](https://www.rfc-editor.org/rfc/rfc8446#section-4.6.1) to the peer. Note that applications may call this at any time after the handshake was completed and as often as they need (up to 2^16 times -- then the ticket nonce depletes).
On any other channel object (i.e. TLS 1.2 server), this throws.

#### `TLS::Server::new_session_ticket_supported()`

Returns `false` if the channel object is not capable of TLS 1.3 session tickets (e.g. because it is a TLS 1.2 server or the TLS 1.3 client did not advertise support for any PSK key exchange mode). If `::send_new_session_ticket()` would likely succeed, this will return `true`.

#### `TLS::Policy::new_session_tickets_upon_handshake_success()`

Once a TLS 1.3 server successfully established a TLS connection with a compatible client, it will automatically send as many session tickets as defined by this policy. This is a convenience function to enable session resumption support without the using application needing to call `::send_new_session_ticket()`. By default the TLS 1.2 implementation offers session resumption, so this returns `1` by default: I.e. clients will receive a single session ticket after a successful handshake.

randombit reviewed (approved):
Approving - just reviewed the last few commits here.

randombit commented:
Your outline for a new session manager interface seems a better match with the realities of how TLS 1.3 does session management. I'm guessing here but it feels like having the implementation notify the session manager that a session was used is a more flexible approach with regards to policy. It also enables LRU style cache pruning.

We may also want to rethink what facilities are offered by the session manager base class itself. In particular right now we do offer out of the box encryption for SQL session managers derived from `Session_Manager_SQL` which seems to me a nice convenience, but if someone need to write say a flatfile session manager they would have to completely reinvent this.

reneme commented:
Re: Session Management revamp

My central idea is to give the `Session_Manager` more control over the session handling. Particularly, it should be in charge of the `Session_ID` and `Session_Ticket`. Hopefully we'll be able to strike a good balance with user convenience, reusability and expected technical understanding to implement application-specific session managers. Providing low-level building blocks (and some safe guards) to facilitate custom implementations is certainly in line with what I have in mind.

I'm currently putting together a working prototype that (roughly) follows the interface I sketched above. Once that is remotely presentable, I'll open a draft PR. Hopefully this will still happen this year. 🤡

reneme commented on src/bogo_shim/config.json:
```diff
@@ -68,30 +62,13 @@
 
         "ExtraClientEncryptedExtension-TLS-TLS13": "TLS 1.3 server UNIMPLEMENTED",
 
+        "Server-Verify-*-TLS13": "TODO: FIXME - missing deny-list for outdated signature schemes",
+        "Server-VerifyDefault-*-TLS13": "TODO: FIXME - missing deny-list for outdated signature schemes",
```
This should be fixed in a follow-up PR. It seems to be an issue in the selection of signature schemes with client auth (?). The test was not enabled before, as it required session resumption

reneme commented:
Note: Lets move the `Session_Manager` discussion over to #3150.

reneme commented:
This is now rebased (finally) onto the `Session_Manager` overhaul and also the various recent enum-class changes. The new API integrated nicely and this should converge soon. Still, I expect some CI breakage and will do another round of self-review and clean-ups.

Nevertheless, this PR chain (#3244, #3150 and #3140) comes together and should be ready for review. Some refactorings or prep-work could probably be moved from here down to #3244 though I'm a bit unsure whether this is worth the head ache.

reneme commented:
Rebased after enum changes prevented a clean merge. Ready for review after #3150 is merged.

reneme commented:
Rebased to master, after #3150 was merged.

reneme commented on src/lib/tls/msg_server_hello.cpp:
```diff
@@ -711,6 +720,8 @@ uint16_t choose_ciphersuite(const Client_Hello_13& ch, const Policy& policy)
 
    for(auto suite_id : pref_list)
       {
+      // TODO: take potentially available PSKs into account to select
+      //       a compatible ciphersuite (if possible).
```
Not necessarily. It would also be beneficial to base the cipher suite selection on potentially available resumable sessions. But that is a somewhat optional optimization that can be done later.

reneme replied:
On a second thought: Maybe we actually want to look into this TODO once again before merging.

reneme replied:
I elaborated this TODO quite a bit more, hoping that future-us will be grateful. For now, the TODO is legitimate, though.

```
      // TODO: take potentially available PSKs into account to select a
      //       compatible ciphersuite.
      //
      // Assuming the client sent one or more PSKs, we would first need to find
      // the hash functions they are associated to. For session tickets, that
      // would mean decrypting the ticket and comparing the cipher suite used in
      // those tickets. For (currently not yet supported) pre-assigned PSKs, the
      // hash function needs to be specified along with them.
      //
      // Then we could refine the ciphersuite selection using the required hash
      // function for the PSK(s) we are wishing to use down the road.
      //
      // For now, we just negotiate the cipher suite blindly and hope for the
      // best. As long as PSKs are used for session resumption only, this has a
      // high chance of success. Previous handshakes with this client have very
      // likely selected the same ciphersuite anyway.
      //
      // See also RFC 8446 4.2.11
      //    When session resumption is the primary use case of PSKs, the most
      //    straightforward way to implement the PSK/cipher suite matching
      //    requirements is to negotiate the cipher suite first [...].
```

reneme commented on src/lib/tls/msg_server_hello.cpp:
```diff
@@ -750,6 +772,34 @@ Server_Hello_13::Server_Hello_13(const Client_Hello_13& ch,
+      const auto psk_modes = ch_exts.get<PSK_Key_Exchange_Modes>();
+      BOTAN_ASSERT_NONNULL(psk_modes);
+
+      // TODO: also support non-DHE PSK Key-Exchange mode
+      if(value_exists(psk_modes->modes(), PSK_Key_Exchange_Mode::PSK_DHE_KE))
```
The comment is misleading, sorry. What I mean is that we should (at some point) also support a PSK-based exchange that does not perform an additional asymmetric ephemeral key exchange. I'll change that.

For reference: TLS 1.3 allow three key exchange modes:

* Ephemeral key only (the common "full" handshake)
* Ephemeral key + PSK (what we currently support for session resumption)
* PSK only (what this TODO requests to be supported) ... this would be a fully symmetric TLS handshake solely based on a PSK (or previous session) (with weaker forward secrecy guarantees)

reneme commented on src/lib/tls/msg_session_ticket.cpp:
```diff
@@ -70,11 +100,21 @@ New_Session_Ticket_13::New_Session_Ticket_13(const std::vector<uint8_t>& buf,
 
    m_extensions.deserialize(reader, from, type());
 
+   // RFC 8446 4.6.1
+   //    The sole extension currently defined for NewSessionTicket is
+   //    "early_data", indicating that the ticket may be used to send 0-RTT
+   //    data [...]. Clients MUST ignore unrecognized extensions.
+   if(m_extensions.contains_implemented_extensions_other_than({Extension_Code::EarlyData}))
+      {
+      throw TLS_Exception(Alert::IllegalParameter,
+                          "NewSessionTicket message contained unexpected extension");
```
That's what it does. `Sessions::contains_other_than()` would be true if _any_ other extensions (including extensions we don't implement) except the ones listed is available. That would not fulfill the "unrecognized" in the RFC. `::contains_implemented_extensions_other_than()` is true if extensions we implement are contained that are not listed. I.e. "unrecognized" additional extensions are okay.

lieser replied:
Ok, seem to have missed the `implemented` part.

reneme commented on src/lib/tls/tls12/tls_server_impl_12.cpp:
```diff
@@ -485,6 +485,7 @@ void Server_Impl_12::process_client_hello_msg(const Handshake_State* active_stat
                        policy(),
                        pending_state.client_hello());
 
+
```
No. thanks.

reneme commented on src/lib/tls/tls13/tls_channel_impl_13.h:
```diff
@@ -58,14 +50,41 @@ class Channel_Impl_13 : public Channel_Impl
+            Handshake_Layer& m_handshake_layer;
+         };
+
+      class AggregatedHandshakeMessages : public AggregatedMessages
```
Updated for improved clarity.

reneme commented on src/lib/tls/tls13/tls_channel_impl_13.h:
```diff
@@ -209,12 +233,23 @@ class Channel_Impl_13 : public Channel_Impl
          }
 
-      void send_post_handshake_message(const Post_Handshake_Message_13 message);
+      std::vector<uint8_t> send_post_handshake_message(Post_Handshake_Message_13 message)
```
I changed it mainly for consistency with `send_handshake_message()`. The return value is not used anywhere in fact. It is just returned from `.send()` anyway and moving it out of this method didn't seem like a performance penalty. I don't have strong feelings either way.

lieser replied:
ok with leaving as is

reneme commented on src/lib/tls/tls13/tls_client_impl_13.cpp:
```diff
@@ -130,12 +130,37 @@ bool Client_Impl_13::handshake_finished() const
+   if(!cert_chain.empty())
+      {
+      const auto& server_cert = cert_chain.front();
+      if(!server_cert.is_self_signed() &&
```
I've gotten this idiom from somewhere else. Frankly I don't remember from where.
Nevertheless, I found it reasonable, as self-signed certs don't come with any security guarantee anyway. But I do agree that redoing this check for a resumption attempt probably isn't breaking anything either.

lieser replied:
I would say we should always do the check, as the RFC does not mention any exception for self signed certs.

reneme replied:
I introduced this check because of the RFC paragraph referenced in the code. The exception for self-signed certificates was (presumably) introduced to pass BoGo tests (that use certificates without SNI/hostname information).

I came to the conclusion that we should not hard-code this check at all.

1. **SNI validation against the original session's certificate**
  Actual validation of the certificate is usually done via `Callbacks::tls_verify_cert_chain()`. In this callback applications can override any SNI restrictions in an ordinary handshake. Therefore, we cannot assume that a hard-coded SNI check for certificates in the resumption case would actually succeed (for instance BoGo tests would fail).
  Instead, we might re-call `tls_verify_cert_chain()` for the stored server cert chain before resumption. However, the RFC does not request a full re-validation of certificates.

2. **Equality of new SNI hostname and the one stored in the session**
  This is practically a given. We are querying the `Session_Manager` for the SNI value. Assuming that the manager does not lie to us, we don't need to re-validate that here.
  
Conclusion: Both checks are out of place here and can be implemented by an application (in a custom session manager) if needed.

reneme commented on src/lib/tls/tls13/tls_client_impl_13.cpp:
```diff
@@ -130,12 +130,37 @@ bool Client_Impl_13::handshake_finished() const
+          X509_Time(callbacks().tls_current_timestamp()) > server_cert.not_after()))
+         { return std::nullopt; }
+      }
+   else if(!session_to_resume.session.server_info().empty() &&
```
Resolving with: https://github.com/randombit/botan/pull/3140#discussion_r1122754582

reneme commented on src/lib/tls/tls13/tls_server_impl_13.cpp:
```diff
@@ -47,11 +47,65 @@ std::string Server_Impl_13::application_protocol() const
+          value_exists(m_handshake_state.client_hello().extensions().get<PSK_Key_Exchange_Modes>()->modes(), PSK_Key_Exchange_Mode::PSK_DHE_KE);
+   }
+
+void Server_Impl_13::send_new_session_tickets(const size_t tickets)
```
Fair point. We can easily collect this information, so why not hand it out.

reneme commented:
Rebased yet again after #3323 caused a conflict. Also fixed [the last bogo test for this PR](https://github.com/randombit/botan/pull/3140/commits/bcd428181feeae9afd9d9069bc27eb3715049670).
